### ISSUE TITLE
#define real_t in generated function header leads to compilation errors if user code typedefs real_t

### ISSUE THREAD

**Issue by mkotlyar** [casadi/casadi#1773]
Let's say we have a CasADi-generated header file with interface to generated C functions. It defines the `real_t` macro (which is essentially a `typedef`):

``` c
// generated.h
#ifndef real_t
#define real_t double
#define to_double(x) (double) x
#define to_int(x) (int) x
#endif /* real_t */
```

Now, if my code does

``` c++
#include "generated.h"

class MyClass {
    typedef double real_t;
};
```

then I get following compilation errors:

```
error: multiple types in one declaration
error: declaration does not declare anything [-fpermissive]
note: in expansion of macro ‘real_t’
```

Obviously, `typedef double real_t;` is expanded to `typedef double double;`, which is a nonsense. Changing `#define real_t double` to `typedef double real_t;` in `generated.h` solves the problem.

I think it is a better idea to explicitly write the floating point type (`double` or `float` or whatever else it can be) in the generated files. This will avoid potential conflicts with names defined in user code.

**Comment by jaeandersson** [comment]
A simple solution is to simply add an "undef real_t` to the generated header. Another solution is to rename the "real_t" macro to something less ambiguous, like `casadi_real_t`.

**Comment by jaeandersson** [comment]
Anyway, in the meantime, I would suggest that you simply "undef" the "real_t":
```c++
#include "generated.h"
#undef real_t

class MyClass {
    typedef double real_t;
};
```

**Comment by mkotlyar** [comment]
`#undef real_t` is exactly what I did. Renaming `real_t` to something less ambitious is also good :-)

**Comment by jaeandersson** [comment]
Fixed in develop - will be included in CasADi 3.3. CasADi now uses the typename `casadi_real` everywhere.
